File: test/cli/setup.test.ts
import { afterEach, describe, expect, test } from "bun:test";
import { lstatSync, readlinkSync, symlinkSync } from "node:fs";
import { chmod, mkdir, mkdtemp, readFile, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import {
  installCliCommands,
  installKeybindings,
  stripDeadBindings,
  readOwnership,
  resolveBinDir,
  resolveHerdrConfigPath,
} from "../../src/cli";
import { PRODUCT_VERSION } from "../../src/context";

const dirs: string[] = [];
afterEach(async () => {
  await Promise.all(dirs.splice(0).map((d) => rm(d, { recursive: true, force: true })));
});

async function writeExecutable(path: string, body: string): Promise<string> {
  await writeFile(path, `#!/bin/sh\n${body}`);
  await chmod(path, 0o755);
  return path;
}

async function writeRecordingHerdr(dir: string): Promise<{ bin: string; log: string }> {
  const log = join(dir, "herdr-argv.log");
  const bin = await writeExecutable(
    join(dir, "fake-herdr"),
    `printf '%s\\n' "$*" >> ${JSON.stringify(log)}
if [ "$1" = config ] && [ "$2" = check ]; then
  echo "config: ok"
  exit 0
fi
if [ "$1" = server ] && [ "$2" = reload-config ]; then
  exit 0
fi
exit 1
`,
  );
  return { bin, log };
}

describe("setup paths", () => {
  test("bin dir prefers XDG_BIN_HOME", () => {
    const custom = join(tmpdir(), "hwf-custom-bin");
    expect(resolveBinDir({ XDG_BIN_HOME: custom })).toBe(custom);
  });

  test("Herdr config prefers HERDR_CONFIG_PATH then XDG path", () => {
    expect(resolveHerdrConfigPath({ HERDR_CONFIG_PATH: "/tmp/c.toml" })).toBe("/tmp/c.toml");
    expect(resolveHerdrConfigPath({ XDG_CONFIG_HOME: "/xdg" })).toBe(
      join("/xdg", "herdr", "config.toml"),
    );
  });
});

describe("cli install", () => {
  test("fresh and repeated setup replaces only owned entries", async () => {
    const root = await mkdtemp(join(tmpdir(), "hwf-setup-cli-"));
    dirs.push(root);
    const binDir = join(root, "bin");
    const pluginBin = join(root, "plugin", "bin");
    await mkdir(pluginBin, { recursive: true });
    const source = join(pluginBin, "herdr-workflows");
    await writeFile(source, "fake-binary\n");

    const first = installCliCommands({
      binDir,
      binary: source,
      ephemeral: false,
    });
    expect(
      first.messages.some(
        (m) => m.includes("install") || m.includes("linked") || m.includes("copied"),
      ),
    ).toBe(true);

    expect(readOwnership(binDir).entries["herdr-workflows"]).toBeDefined();
    expect(readOwnership(binDir).entries.hwf).toBeDefined();

    await writeFile(source, "fake-binary-v2\n");
    const second = installCliCommands({
      binDir,
      binary: source,
      ephemeral: false,
    });
    expect(second.messages.length).toBeGreaterThan(0);
    expect(readOwnership(binDir).entries["herdr-workflows"]).toEqual({
      kind: "symlink",
      version: PRODUCT_VERSION,
      source,
    });
  });

  test("foreign entry is preserved and named", async () => {
    const root = await mkdtemp(join(tmpdir(), "hwf-setup-foreign-"));
    dirs.push(root);
    const binDir = join(root, "bin");
    await mkdir(binDir, { recursive: true });
    const foreign = join(binDir, "herdr-workflows");
    await writeFile(foreign, "not-ours\n");
    const source = join(root, "source-bin");
    await writeFile(source, "ours\n");

    const result = installCliCommands({
      binDir,
      binary: source,
      ephemeral: false,
    });
    expect(result.messages.some((m) => m.includes("not owned"))).toBe(true);
    expect(await readFile(foreign, "utf8")).toBe("not-ours\n");
  });

  test("retargeted owned symlink is preserved as foreign", async () => {
    const root = await mkdtemp(join(tmpdir(), "hwf-setup-retargeted-"));
    dirs.push(root);
    const binDir = join(root, "bin");
    const source = join(root, "source-bin");
    const foreign = join(root, "foreign-bin");
    await writeFile(source, "ours\n");
    await writeFile(foreign, "foreign\n");

    installCliCommands({ binDir, binary: source, ephemeral: false });
    const hwf = join(binDir, "hwf");
    await rm(hwf);
    symlinkSync(foreign, hwf);

    const result = installCliCommands({ binDir, binary: source, ephemeral: false });
    expect(result.messages).toContain(
      `skipped cli install: ${hwf} exists and is not owned by herdr-workflows`,
    );
    expect(await readFile(hwf, "utf8")).toBe("foreign\n");
  });

  test("ephemeral setup copies the binary once and symlinks hwf to it", async () => {
    const root = await mkdtemp(join(tmpdir(), "hwf-setup-single-copy-"));
    dirs.push(root);
    const binDir = join(root, "bin");
    const pluginBin = join(root, "plugin", "bin");
    await mkdir(pluginBin, { recursive: true });
    const source = join(pluginBin, "herdr-workflows");
    await writeFile(source, "fake-binary\n");

    installCliCommands({ binDir, binary: source, ephemeral: true });

    const primary = join(binDir, "herdr-workflows");
    const hwf = join(binDir, "hwf");
    expect(lstatSync(primary).isSymbolicLink()).toBe(false);
    expect(lstatSync(hwf).isSymbolicLink()).toBe(true);
    expect(readlinkSync(hwf)).toBe(primary);
    expect(await readFile(hwf, "utf8")).toBe("fake-binary\n");
    expect(readOwnership(binDir).entries.hwf?.kind).toBe("symlink");
  });

  test("ephemeral setup with foreign primary copies hwf instead of dangling", async () => {
    const root = await mkdtemp(join(tmpdir(), "hwf-setup-foreign-primary-"));
    dirs.push(root);
    const binDir = join(root, "bin");
    await mkdir(binDir, { recursive: true });
    await writeFile(join(binDir, "herdr-workflows"), "not-ours\n");
    const pluginBin = join(root, "plugin", "bin");
    await mkdir(pluginBin, { recursive: true });
    const source = join(pluginBin, "herdr-workflows");
    await writeFile(source, "fake-binary\n");

    const result = installCliCommands({ binDir, binary: source, ephemeral: true });

    expect(result.messages.some((m) => m.includes("not owned"))).toBe(true);
    const hwf = join(binDir, "hwf");
    expect(lstatSync(hwf).isSymbolicLink()).toBe(false);
    expect(await readFile(hwf, "utf8")).toBe("fake-binary\n");
  });

  test("ephemeral setup preserves foreign files and symlinks", async () => {
    const root = await mkdtemp(join(tmpdir(), "hwf-setup-ephemeral-"));
    dirs.push(root);
    const binDir = join(root, "bin");
    await mkdir(binDir, { recursive: true });
    const foreignFile = join(binDir, "herdr-workflows");
    const foreignLink = join(binDir, "hwf");
    const elsewhere = join(root, "foreign-target");
    await writeFile(foreignFile, "foreign-file\n");
    await writeFile(elsewhere, "foreign-link-target\n");
    symlinkSync(elsewhere, foreignLink);
    const source = join(root, "source-bin");
    await writeFile(source, "ours\n");

    const result = installCliCommands({
      binDir,
      binary: source,
      ephemeral: true,
    });
    expect(result.messages.filter((m) => m.includes("not owned"))).toHaveLength(2);
    expect(await readFile(foreignFile, "utf8")).toBe("foreign-file\n");
    expect(await readFile(foreignLink, "utf8")).toBe("foreign-link-target\n");
  });

  test("PATH warning is produced by callers when bin dir is absent", () => {
    const missing = resolveBinDir({ XDG_BIN_HOME: join(tmpdir(), "hwf-missing-bin-xyz") });
    expect(missing).toContain("hwf-missing-bin-xyz");
  });
});

describe("keybinding install", () => {
  test("prefix+k launch binding is idempotent and calls config check + reload", async () => {
    const dir = await mkdtemp(join(tmpdir(), "herdr-workflows-keys-bin-"));
    dirs.push(dir);
    const { bin: herdrBin, log: logPath } = await writeRecordingHerdr(dir);
    const path = join(dir, "config.toml");
    await writeFile(path, "");
    await writeFile(logPath, "");

    const first = installKeybindings({
      env: {
        ...process.env,
        HERDR_CONFIG_PATH: path,
        HERDR_BIN_PATH: herdrBin,
      },
    });
    expect(first.messages.join("\n")).toContain("herdr-workflows.launch");
    const text = await readFile(path, "utf8");
    expect(text).toContain("herdr-workflows.launch");
    expect(text).toContain('key = "prefix+k"');
    expect(text).not.toContain("herdr-workflows.results");
    const log = await readFile(logPath, "utf8");
    expect(log).toContain("config check");
    expect(log).toContain("server reload-config");
    expect(first.messages.join("\n")).toContain("herdr reloaded config");

    await writeFile(logPath, "");
    const again = installKeybindings({
      env: {
        ...process.env,
        HERDR_CONFIG_PATH: path,
        HERDR_BIN_PATH: herdrBin,
      },
    });
    expect(again.messages.join("\n")).toContain("already present");
    expect(await readFile(path, "utf8")).toBe(text);
    expect((text.match(/herdr-workflows\.launch/g) ?? []).length).toBe(1);
    expect(await readFile(logPath, "utf8")).toBe("");
  });

  test("strips retired results and legacy kagan/lembas launch", async () => {
    const dir = await mkdtemp(join(tmpdir(), "herdr-workflows-keys-bin-"));
    dirs.push(dir);
    const { bin: herdrBin, log: logPath } = await writeRecordingHerdr(dir);
    const path = join(dir, "config.toml");
    const stale = `# keep me

[[keys.command]]
key = "prefix+k"
type = "plugin_action"
command = "kagan.launch"
description = "launch a kagan workflow (picker)"

[[keys.command]]
key = "prefix+l"
type = "plugin_action"
command = "lembas.launch"
description = "launch lembas"

[[keys.command]]
key = "prefix+r"
type = "plugin_action"
command = "herdr-workflows.results"
description = "view completed herdr-workflows job results"
`;
    await writeFile(path, stale);
    await writeFile(logPath, "");
    const result = installKeybindings({
      env: {
        ...process.env,
        HERDR_CONFIG_PATH: path,
        HERDR_BIN_PATH: herdrBin,
      },
    });
    expect(result.messages.join("\n")).toContain("removed dead");
    const text = await readFile(path, "utf8");
    expect(text).toContain("herdr-workflows.launch");
    expect(text).not.toContain("kagan.launch");
    expect(text).not.toContain("lembas.launch");
    expect(text).not.toContain("herdr-workflows.results");
    expect(text).not.toContain("prefix+r");

    const again = installKeybindings({
      env: {
        ...process.env,
        HERDR_CONFIG_PATH: path,
        HERDR_BIN_PATH: herdrBin,
      },
    });
    expect(again.messages.join("\n")).toContain("already present");
    expect(await readFile(path, "utf8")).toBe(text);
    expect(await readFile(`${path}.hwf.bak`, "utf8")).toBe(stale);
  });

  test("missing herdr validator does not write config", async () => {
    const dir = await mkdtemp(join(tmpdir(), "herdr-workflows-keys-miss-"));
    dirs.push(dir);
    const path = join(dir, "config.toml");
    const original = "# keep me\n";
    await writeFile(path, original);
    const result = installKeybindings({
      env: {
        ...process.env,
        HERDR_CONFIG_PATH: path,
        HERDR_BIN_PATH: join(dir, "missing-herdr"),
      },
      reload: false,
    });
    expect(result.messages.join("\n")).toContain("config check failed");
    expect(await readFile(path, "utf8")).toBe(original);
    expect(await Bun.file(`${path}.hwf.tmp`).exists()).toBe(false);
  });

  test("stripDeadBindings removes only retired tables", () => {
    const text = `
[[keys.command]]
command = "keep.me"

[[keys.command]]
command = "herdr-workflows.results"
`;
    const cleaned = stripDeadBindings(text);
    expect(cleaned).toContain("keep.me");
    expect(cleaned).not.toContain("herdr-workflows.results");
  });

  test("reload-config failure is reported without claiming the running Herdr loaded the binding", async () => {
    const dir = await mkdtemp(join(tmpdir(), "herdr-workflows-keys-reload-"));
    dirs.push(dir);
    const log = join(dir, "herdr-argv.log");
    const herdrBin = await writeExecutable(
      join(dir, "fake-herdr"),
      `printf '%s\\n' "$*" >> ${JSON.stringify(log)}
if [ "$1" = config ] && [ "$2" = check ]; then
  echo "config: ok"
  exit 0
fi
if [ "$1" = server ] && [ "$2" = reload-config ]; then
  echo "reload denied" >&2
  exit 3
fi
exit 1
`,
    );
    const path = join(dir, "config.toml");
    await writeFile(path, "");
    const result = installKeybindings({
      env: {
        ...process.env,
        HERDR_CONFIG_PATH: path,
        HERDR_BIN_PATH: herdrBin,
      },
    });
    const joined = result.messages.join("\n");
    expect(joined).toContain("herdr-workflows.launch");
    expect(joined).toContain("reload-config failed");
    expect(joined).toContain("may not have loaded the binding");
    expect(joined).not.toContain("herdr reloaded config");
    expect(await readFile(path, "utf8")).toContain("herdr-workflows.launch");
  });
});

File: internal/cli/setup_test.go
package cli

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/aorumbayev/herdr-workflows/internal/config"
)

func writeExecutable(t *testing.T, path, body string) {
	t.Helper()
	content := "#!/bin/sh\n" + body
	if err := os.WriteFile(path, []byte(content), 0o755); err != nil {
		t.Fatal(err)
	}
}

func writeRecordingHerdr(t *testing.T, dir string) (bin, log string) {
	t.Helper()
	log = filepath.Join(dir, "herdr-argv.log")
	bin = filepath.Join(dir, "fake-herdr")
	writeExecutable(t, bin, `printf '%s\n' "$*" >> "`+log+`"
if [ "$1" = config ] && [ "$2" = check ]; then
  echo "config: ok"
  exit 0
fi
if [ "$1" = server ] && [ "$2" = reload-config ]; then
  exit 0
fi
exit 1
`)
	return bin, log
}

func TestResolveBinDirPrefersXDGBinHome(t *testing.T) {
	custom := filepath.Join(t.TempDir(), "hwf-custom-bin")
	got := ResolveBinDir(func(string) string { return custom })
	if got != custom {
		t.Fatalf("ResolveBinDir() = %q, want %q", got, custom)
	}
}

func TestResolveHerdrConfigPath(t *testing.T) {
	if got := ResolveHerdrConfigPath(func(k string) string {
		if k == "HERDR_CONFIG_PATH" {
			return "/tmp/c.toml"
		}
		return ""
	}); got != "/tmp/c.toml" {
		t.Fatalf("HERDR_CONFIG_PATH = %q", got)
	}
	want := filepath.Join("/xdg", "herdr", "config.toml")
	if got := ResolveHerdrConfigPath(func(k string) string {
		if k == "XDG_CONFIG_HOME" {
			return "/xdg"
		}
		return ""
	}); got != want {
		t.Fatalf("XDG path = %q, want %q", got, want)
	}
}

func TestInstallCliCommandsFreshAndRepeated(t *testing.T) {
	root := t.TempDir()
	binDir := filepath.Join(root, "bin")
	pluginBin := filepath.Join(root, "plugin", "bin")
	if err := os.MkdirAll(pluginBin, 0o755); err != nil {
		t.Fatal(err)
	}
	source := filepath.Join(pluginBin, "herdr-workflows")
	if err := os.WriteFile(source, []byte("fake-binary\n"), 0o644); err != nil {
		t.Fatal(err)
	}

	first := InstallCliCommands(binDir, source, false)
	if !messagesContainInstall(first.Messages) {
		t.Fatalf("first messages = %v", first.Messages)
	}
	reg := ReadOwnership(binDir)
	if reg.Entries["herdr-workflows"].Kind != ownedSymlink {
		t.Fatalf("herdr-workflows entry = %+v", reg.Entries["herdr-workflows"])
	}
	if reg.Entries["hwf"].Kind == "" {
		t.Fatal("missing hwf entry")
	}

	if err := os.WriteFile(source, []byte("fake-binary-v2\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	second := InstallCliCommands(binDir, source, false)
	if len(second.Messages) == 0 {
		t.Fatal("expected messages on repeat install")
	}
	reg = ReadOwnership(binDir)
	absSource, _ := filepath.Abs(source)
	entry := reg.Entries["herdr-workflows"]
	if entry.Kind != ownedSymlink || entry.Version != config.ProductVersion || entry.Source != absSource {
		t.Fatalf("entry = %+v, want symlink %s %s", entry, config.ProductVersion, absSource)
	}
}

func messagesContainInstall(msgs []string) bool {
	for _, m := range msgs {
		if strings.Contains(m, "install") || strings.Contains(m, "linked") || strings.Contains(m, "copied") {
			return true
		}
	}
	return false
}

func TestInstallCliCommandsPreservesForeignEntry(t *testing.T) {
	root := t.TempDir()
	binDir := filepath.Join(root, "bin")
	if err := os.MkdirAll(binDir, 0o755); err != nil {
		t.Fatal(err)
	}
	foreign := filepath.Join(binDir, "herdr-workflows")
	if err := os.WriteFile(foreign, []byte("not-ours\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	source := filepath.Join(root, "source-bin")
	if err := os.WriteFile(source, []byte("ours\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	result := InstallCliCommands(binDir, source, false)
	if !containsMessage(result.Messages, "not owned") {
		t.Fatalf("messages = %v", result.Messages)
	}
	data, err := os.ReadFile(foreign)
	if err != nil || string(data) != "not-ours\n" {
		t.Fatalf("foreign file = %q err=%v", data, err)
	}
}

func TestInstallCliCommandsRetargetedOwnedSymlink(t *testing.T) {
	root := t.TempDir()
	binDir := filepath.Join(root, "bin")
	source := filepath.Join(root, "source-bin")
	foreign := filepath.Join(root, "foreign-bin")
	for _, p := range []struct {
		path, body string
	}{
		{source, "ours\n"},
		{foreign, "foreign\n"},
	} {
		if err := os.WriteFile(p.path, []byte(p.body), 0o644); err != nil {
			t.Fatal(err)
		}
	}
	InstallCliCommands(binDir, source, false)
	hwf := filepath.Join(binDir, "hwf")
	if err := os.Remove(hwf); err != nil {
		t.Fatal(err)
	}
	if err := os.Symlink(foreign, hwf); err != nil {
		t.Fatal(err)
	}
	result := InstallCliCommands(binDir, source, false)
	want := "skipped cli install: " + hwf + " exists and is not owned by herdr-workflows"
	if !containsMessage(result.Messages, want) {
		t.Fatalf("messages = %v", result.Messages)
	}
	data, err := os.ReadFile(hwf)
	if err != nil || string(data) != "foreign\n" {
		t.Fatalf("hwf = %q err=%v", data, err)
	}
}

func TestInstallCliCommandsEphemeralSingleCopy(t *testing.T) {
	root := t.TempDir()
	binDir := filepath.Join(root, "bin")
	pluginBin := filepath.Join(root, "plugin", "bin")
	if err := os.MkdirAll(pluginBin, 0o755); err != nil {
		t.Fatal(err)
	}
	source := filepath.Join(pluginBin, "herdr-workflows")
	if err := os.WriteFile(source, []byte("fake-binary\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	InstallCliCommands(binDir, source, true)

	primary := filepath.Join(binDir, "herdr-workflows")
	hwf := filepath.Join(binDir, "hwf")
	pst, err := os.Lstat(primary)
	if err != nil {
		t.Fatal(err)
	}
	if pst.Mode()&os.ModeSymlink != 0 {
		t.Fatal("primary should be a copy, not symlink")
	}
	hst, err := os.Lstat(hwf)
	if err != nil {
		t.Fatal(err)
	}
	if hst.Mode()&os.ModeSymlink == 0 {
		t.Fatal("hwf should be symlink")
	}
	link, err := os.Readlink(hwf)
	if err != nil || link != primary {
		t.Fatalf("hwf link = %q err=%v", link, err)
	}
	data, err := os.ReadFile(hwf)
	if err != nil || string(data) != "fake-binary\n" {
		t.Fatalf("hwf content = %q err=%v", data, err)
	}
	if ReadOwnership(binDir).Entries["hwf"].Kind != ownedSymlink {
		t.Fatal("hwf should be recorded as symlink")
	}
}

func TestInstallCliCommandsEphemeralForeignPrimary(t *testing.T) {
	root := t.TempDir()
	binDir := filepath.Join(root, "bin")
	if err := os.MkdirAll(binDir, 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(binDir, "herdr-workflows"), []byte("not-ours\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	pluginBin := filepath.Join(root, "plugin", "bin")
	if err := os.MkdirAll(pluginBin, 0o755); err != nil {
		t.Fatal(err)
	}
	source := filepath.Join(pluginBin, "herdr-workflows")
	if err := os.WriteFile(source, []byte("fake-binary\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	result := InstallCliCommands(binDir, source, true)
	if !containsMessage(result.Messages, "not owned") {
		t.Fatalf("messages = %v", result.Messages)
	}
	hwf := filepath.Join(binDir, "hwf")
	st, err := os.Lstat(hwf)
	if err != nil {
		t.Fatal(err)
	}
	if st.Mode()&os.ModeSymlink != 0 {
		t.Fatal("hwf should be copied when primary is foreign")
	}
	data, err := os.ReadFile(hwf)
	if err != nil || string(data) != "fake-binary\n" {
		t.Fatalf("hwf = %q err=%v", data, err)
	}
}

func TestInstallCliCommandsEphemeralPreservesForeign(t *testing.T) {
	root := t.TempDir()
	binDir := filepath.Join(root, "bin")
	if err := os.MkdirAll(binDir, 0o755); err != nil {
		t.Fatal(err)
	}
	foreignFile := filepath.Join(binDir, "herdr-workflows")
	foreignLink := filepath.Join(binDir, "hwf")
	elsewhere := filepath.Join(root, "foreign-target")
	for _, p := range []struct {
		path, body string
	}{
		{foreignFile, "foreign-file\n"},
		{elsewhere, "foreign-link-target\n"},
	} {
		if err := os.WriteFile(p.path, []byte(p.body), 0o644); err != nil {
			t.Fatal(err)
		}
	}
	if err := os.Symlink(elsewhere, foreignLink); err != nil {
		t.Fatal(err)
	}
	source := filepath.Join(root, "source-bin")
	if err := os.WriteFile(source, []byte("ours\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	result := InstallCliCommands(binDir, source, true)
	notOwned := 0
	for _, m := range result.Messages {
		if strings.Contains(m, "not owned") {
			notOwned++
		}
	}
	if notOwned != 2 {
		t.Fatalf("not owned count = %d, messages = %v", notOwned, result.Messages)
	}
	for _, pair := range []struct {
		path, want string
	}{
		{foreignFile, "foreign-file\n"},
		{foreignLink, "foreign-link-target\n"},
	} {
		data, err := os.ReadFile(pair.path)
		if err != nil || string(data) != pair.want {
			t.Fatalf("%s = %q err=%v", pair.path, data, err)
		}
	}
}

func containsMessage(msgs []string, want string) bool {
	for _, m := range msgs {
		if m == want || strings.Contains(m, want) {
			return true
		}
	}
	return false
}

func TestInstallKeybindingsIdempotent(t *testing.T) {
	dir := t.TempDir()
	herdrBin, logPath := writeRecordingHerdr(t, dir)
	path := filepath.Join(dir, "config.toml")
	if err := os.WriteFile(path, nil, 0o644); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(logPath, nil, 0o644); err != nil {
		t.Fatal(err)
	}
	getenv := func(k string) string {
		switch k {
		case "HERDR_CONFIG_PATH":
			return path
		case "HERDR_BIN_PATH":
			return herdrBin
		}
		return os.Getenv(k)
	}
	first := InstallKeybindings(KeybindingInstallOpts{Getenv: getenv})
	joined := strings.Join(first.Messages, "\n")
	if !strings.Contains(joined, "herdr-workflows.launch") {
		t.Fatalf("messages = %q", joined)
	}
	text, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	body := string(text)
	for _, want := range []string{"herdr-workflows.launch", `key = "prefix+k"`} {
		if !strings.Contains(body, want) {
			t.Fatalf("config missing %q: %q", want, body)
		}
	}
	if strings.Contains(body, "herdr-workflows.results") {
		t.Fatalf("config should not contain results binding: %q", body)
	}
	log, err := os.ReadFile(logPath)
	if err != nil {
		t.Fatal(err)
	}
	logText := string(log)
	if !strings.Contains(logText, "config check") || !strings.Contains(logText, "server reload-config") {
		t.Fatalf("log = %q", logText)
	}
	if !strings.Contains(joined, "herdr reloaded config") {
		t.Fatalf("messages = %q", joined)
	}

	if err := os.WriteFile(logPath, nil, 0o644); err != nil {
		t.Fatal(err)
	}
	again := InstallKeybindings(KeybindingInstallOpts{Getenv: getenv})
	if !strings.Contains(strings.Join(again.Messages, "\n"), "already present") {
		t.Fatalf("again messages = %v", again.Messages)
	}
	text2, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	if string(text2) != body {
		t.Fatal("config changed on second install")
	}
	if strings.Count(body, "herdr-workflows.launch") != 1 {
		t.Fatalf("launch count = %d", strings.Count(body, "herdr-workflows.launch"))
	}
	log2, err := os.ReadFile(logPath)
	if err != nil || len(log2) != 0 {
		t.Fatalf("log after idempotent = %q err=%v", log2, err)
	}
}

func TestInstallKeybindingsStripsRetired(t *testing.T) {
	dir := t.TempDir()
	herdrBin, logPath := writeRecordingHerdr(t, dir)
	path := filepath.Join(dir, "config.toml")
	stale := `# keep me

[[keys.command]]
key = "prefix+k"
type = "plugin_action"
command = "kagan.launch"
description = "launch a kagan workflow (picker)"

[[keys.command]]
key = "prefix+l"
type = "plugin_action"
command = "lembas.launch"
description = "launch lembas"

[[keys.command]]
key = "prefix+r"
type = "plugin_action"
command = "herdr-workflows.results"
description = "view completed herdr-workflows job results"
`
	if err := os.WriteFile(path, []byte(stale), 0o644); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(logPath, nil, 0o644); err != nil {
		t.Fatal(err)
	}
	getenv := func(k string) string {
		switch k {
		case "HERDR_CONFIG_PATH":
			return path
		case "HERDR_BIN_PATH":
			return herdrBin
		}
		return os.Getenv(k)
	}
	result := InstallKeybindings(KeybindingInstallOpts{Getenv: getenv})
	if !strings.Contains(strings.Join(result.Messages, "\n"), "removed dead") {
		t.Fatalf("messages = %v", result.Messages)
	}
	text, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	body := string(text)
	for _, forbidden := range []string{"kagan.launch", "lembas.launch", "herdr-workflows.results", "prefix+r"} {
		if strings.Contains(body, forbidden) {
			t.Fatalf("config still contains %q: %q", forbidden, body)
		}
	}
	if !strings.Contains(body, "herdr-workflows.launch") {
		t.Fatalf("missing launch binding: %q", body)
	}
	bak, err := os.ReadFile(path + ".hwf.bak")
	if err != nil || string(bak) != stale {
		t.Fatalf("backup = %q err=%v", bak, err)
	}
	again := InstallKeybindings(KeybindingInstallOpts{Getenv: getenv})
	if !strings.Contains(strings.Join(again.Messages, "\n"), "already present") {
		t.Fatalf("again = %v", again.Messages)
	}
	text2, err := os.ReadFile(path)
	if err != nil || string(text2) != body {
		t.Fatal("config changed on repeat")
	}
}

func TestStripDeadBindings(t *testing.T) {
	text := `
[[keys.command]]
command = "keep.me"

[[keys.command]]
command = "herdr-workflows.results"
`
	cleaned := StripDeadBindings(text)
	if !strings.Contains(cleaned, "keep.me") {
		t.Fatalf("cleaned = %q", cleaned)
	}
	if strings.Contains(cleaned, "herdr-workflows.results") {
		t.Fatalf("cleaned = %q", cleaned)
	}
}

func TestInstallKeybindingsMissingValidator(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "config.toml")
	original := "# keep me\n"
	if err := os.WriteFile(path, []byte(original), 0o644); err != nil {
		t.Fatal(err)
	}
	reload := false
	result := InstallKeybindings(KeybindingInstallOpts{
		Getenv: func(k string) string {
			switch k {
			case "HERDR_CONFIG_PATH":
				return path
			case "HERDR_BIN_PATH":
				return filepath.Join(dir, "missing-herdr")
			}
			return os.Getenv(k)
		},
		Reload: &reload,
	})
	if !strings.Contains(strings.Join(result.Messages, "\n"), "config check failed") {
		t.Fatalf("messages = %v", result.Messages)
	}
	data, err := os.ReadFile(path)
	if err != nil || string(data) != original {
		t.Fatalf("config = %q err=%v", data, err)
	}
	if _, err := os.Stat(path + ".hwf.tmp"); !os.IsNotExist(err) {
		t.Fatalf("tmp file should not exist: %v", err)
	}
}

func TestInstallKeybindingsReloadFailure(t *testing.T) {
	dir := t.TempDir()
	log := filepath.Join(dir, "herdr-argv.log")
	herdrBin := filepath.Join(dir, "fake-herdr")
	writeExecutable(t, herdrBin, `printf '%s\n' "$*" >> "`+log+`"
if [ "$1" = config ] && [ "$2" = check ]; then
  echo "config: ok"
  exit 0
fi
if [ "$1" = server ] && [ "$2" = reload-config ]; then
  echo "reload denied" >&2
  exit 3
fi
exit 1
`)
	path := filepath.Join(dir, "config.toml")
	if err := os.WriteFile(path, nil, 0o644); err != nil {
		t.Fatal(err)
	}
	getenv := func(k string) string {
		switch k {
		case "HERDR_CONFIG_PATH":
			return path
		case "HERDR_BIN_PATH":
			return herdrBin
		}
		return os.Getenv(k)
	}
	result := InstallKeybindings(KeybindingInstallOpts{Getenv: getenv})
	joined := strings.Join(result.Messages, "\n")
	for _, want := range []string{"herdr-workflows.launch", "reload-config failed", "may not have loaded the binding"} {
		if !strings.Contains(joined, want) {
			t.Fatalf("messages missing %q: %q", want, joined)
		}
	}
	if strings.Contains(joined, "herdr reloaded config") {
		t.Fatalf("messages = %q", joined)
	}
	body, err := os.ReadFile(path)
	if err != nil || !strings.Contains(string(body), "herdr-workflows.launch") {
		t.Fatalf("config = %q err=%v", body, err)
	}
}

func TestResolveBinDirPathWarningShape(t *testing.T) {
	missing := ResolveBinDir(func(k string) string {
		if k == "XDG_BIN_HOME" {
			return filepath.Join(t.TempDir(), "hwf-missing-bin-xyz")
		}
		return ""
	})
	if !strings.Contains(missing, "hwf-missing-bin-xyz") {
		t.Fatalf("bin dir = %q", missing)
	}
}
